arangod/Aql/Functions.cpp

```diff
@@ -6507,3 +6518,412 @@ AqlValue Functions::Fail(arangodb::aql::Query* query, transaction::Methods* trx,
+      TRI_ASSERT(part <= 4);
+      return std::to_string(part);
+    }},
+  {"%a", [](std::string& wrk, tp_sys_clock_ms const& tp) {
+      auto ymd = year_month_day{floor<days>(tp)};
+      auto lastMonthDay = ymd.year() / ymd.month() / last;
+      return std::to_string(static_cast<uint64_t>(unsigned(lastMonthDay.day())));
+    }},
+  {"%%", [](std::string& wrk, tp_sys_clock_ms const& tp) { return std::string("%"); }},
+  {"%", [](std::string& wrk, tp_sys_clock_ms const& tp) { return ""; }}
+};
+
+std::string regex_replace(std::string const& search,
+                          const std::regex& re,
+                          dateMapType const& dateMap,
+                          tp_sys_clock_ms const& tp)
+{
+  std::string s;
+
+  auto first = search.begin();
+  auto last = search.end();
+  typename std::smatch::difference_type
+    positionOfLastMatch = 0;
+  auto endOfLastMatch = first;
+
+  auto callback = [&tp, &endOfLastMatch, &positionOfLastMatch, &s, &dateMap](const std::smatch& match)
+    {
+      auto positionOfThisMatch = match.position(0);
+      auto diff = positionOfThisMatch - positionOfLastMatch;
+
+      auto startOfThisMatch = endOfLastMatch;
+      std::advance(startOfThisMatch, diff);
+
+      s.append(endOfLastMatch, startOfThisMatch);
+      auto got = dateMap.find(match.str(0));
+      if (got != dateMap.end()) {
+        std::string sanotehu;
+        s.append(got->second(sanotehu, tp));
+      }
+      auto lengthOfMatch = match.length(0);
+
+      positionOfLastMatch = positionOfThisMatch + lengthOfMatch;
+
+      endOfLastMatch = startOfThisMatch;
+      std::advance(endOfLastMatch, lengthOfMatch);
+    };
+
+  std::regex_iterator<std::string::const_iterator> end; /// TODO - howto initialize?
+  std::regex_iterator<std::string::const_iterator> begin(first, last, re);
+  std::for_each(begin, end, callback);
+
+  s.append(endOfLastMatch, last);
+
+  return s;
+}
+std::regex theDateFormatRegex;
+
+static std::string dateFormatRegex(const std::string& s, tp_sys_clock_ms& tp)
+{
+  return regex_replace(s, theDateFormatRegex, dateMap, tp);
+}
+
+void Functions::init() {
+  std::string myregex;
+
+  dateMap.reserve(sortedDateMap.size());
+  std::for_each(sortedDateMap.begin(), sortedDateMap.end(), [&myregex](std::pair<std::string, format_func_t> p) {
+      (myregex.length() > 0) ? myregex += "|" + p.first : myregex = p.first;
+      dateMap.insert(std::make_pair(p.first, p.second));
+    });
+  theDateFormatRegex = std::regex(myregex);
+}
+  
+
+
+/// @brief function DATE_FORMAT
+AqlValue Functions::DateFormat(arangodb::aql::Query* query,
+                               transaction::Methods* trx,
+                               VPackFunctionParameters const& params)
+{
+  static char const* AFN = "DATE_FORMAT";
+  tp_sys_clock_ms tp;
+  ValidateParameters(params, AFN, 2, 2);
+
+  if (!ParameterToTimePoint(query, trx, params, tp, AFN, 0)) {
+    return AqlValue(AqlValueHintNull());
```

Comment:
will we add a warning here so the end user knows the time value passed is invalid?